Update README for fluid telemetry package

rohandubal opened the pull request:
## Description

Update README to better align with (future) documentation that is planned. 

> Note: One pending todo to link queries for this telemetry, which will be coming in shortly. Getting this PR started to gather initial feedback.

rohandubal committed:
Update README for fluid telemetry package

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -2,21 +2,29 @@
 This package contains code enabling the production and consumption of typed telemetry for Fluid Framework applications. The typed telemetry from this package is used as the backbone for different Fluid Framework cloud offerings such as dashboards and alarms for Fluid applications. This package can also be used as a reference for customizing and creating your own telemetry solution if desired.
 
-At this time, the package enables collection of Fluid Container related telemetry. In the future more areas may be added as needed by customers.
+At this time, the package enables collection of Fluid Container related telemetry. In the future, more areas would be added as needs evolve.
```
```suggestion
At this time, the package enables collection of telemetry related to the Fluid Container only. In the future, more areas would be added as needs evolve.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -62,17 +73,26 @@ startTelemetry(telemetryConfig);
+## Use Case 2: Setup a custom telemetry consumer
 
-In this example, you'll walk through the basic setup process to start getting container telemetry to be produced and logging it to the console.
+To use a custom telemetry consumer and integrate with a cloud provider or custom dashboard solution, you need to extend the ITelemetryConsumer package.
```
```suggestion
To use a custom telemetry consumer and integrate with a cloud provider or custom dashboard solution, you need to implement the `ITelemetryConsumer` interface.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -62,17 +73,26 @@ startTelemetry(telemetryConfig);
-### Step 2: First, we'll have to create our own telemetry consumer which extends the ITelemetryConsumer interface. Let's look at an example that will simply console.log the telemetry.
+### Step 2: Setup a custom telemetry consumer
+
+Here we'll create our own telemetry consumer which extends the ITelemetryConsumer interface. Let's look at an example that will simply log the telemetry into console.
```
```suggestion
Here we'll create our own telemetry consumer which extends the `ITelemetryConsumer` interface. Let's look at an example that will simply log the telemetry to console.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -62,17 +73,26 @@ startTelemetry(telemetryConfig);
 -   Using NPM: `npm install @fluidframework/fluid-telemetry`
 
-### Step 2: First, we'll have to create our own telemetry consumer which extends the ITelemetryConsumer interface. Let's look at an example that will simply console.log the telemetry.
+### Step 2: Setup a custom telemetry consumer
```
```suggestion
### Step 2: Define a custom telemetry consumer
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
````diff
@@ -84,7 +104,9 @@ class MySimpleTelemetryConsumer implements ITelemetryConsumer {
 ```
 
-### Step 3: Now, let's start the telemetry production and hook in our telemetry consumer from step 2. We will be initializing our telemetry collection where we initialize our containers:
+### Step 3: Setup the custom telemetry consumer
````
```suggestion
### Step 3: Setup telemetry collection
```

rohandubal committed:
Update packages/framework/client-logger/fluid-telemetry/README.md

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

rohandubal committed:
Update packages/framework/client-logger/fluid-telemetry/README.md

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

rohandubal committed:
Update packages/framework/client-logger/fluid-telemetry/README.md

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

rohandubal committed:
Update packages/framework/client-logger/fluid-telemetry/README.md

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

rohandubal committed:
Update packages/framework/client-logger/fluid-telemetry/README.md

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

rohandubal committed:
Add docs around running telemetry queries

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -75,8 +75,7 @@ startTelemetry(telemetryConfig);
 
-You can find these queries [here]().
-// TODO: rohandubal:Fix me
+You can find these queries [here](#telemetry_visualization).
```
Nit, I was told once that for screen readers it is useful to have more descriptive text in links.
 ```suggestion
You can find these queries in [the "Interpreting telemetry data" section](#telemetry_visualization) below.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
\ No newline at end of file
+You can now run the app and see the telemetry being printed on your console.
+
+## <a name="telemetry_visualization">Interpreting telemetry data</a>
```
Oh, interesting, is this a standard thing with markdown? I would have used `[link text](#interpreting-telemetry-data)` above but fixing an id seems way better.

rohandubal replied:
yup! I learnt it today also. Was not aware previously; this comes in handy.

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+### Accessing App Insights Portal
+
+Before we can query, we must first navigate to your Azure App Insights telemetry page. To do this, go to your Azure App Insights Instance and Click on the “Logs” tab under Monitoring
```
```suggestion
Before we can query, we must first navigate to your Azure App Insights telemetry page. To do this, go to your Azure App Insights Instance and click on the “Logs” tab under Monitoring.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+![Logs on App Insights Portal](readme_images/telemetry_1.png)
+
+Now, close out the “Queries” pane if it showed up for you and you will be in the view where we can now execute our queries. Note that if you are using the Fluid Azure App Insights logger, your telemetry data will be available in the “customEvents” table. 
```
Should this make reference to this package specifically instead of the "Fluid Azure App Insights Logger"? Or is it intentionally referring to the _other_ package, that pushes the internal telemetry?
```suggestion
Now, close out the “Queries” pane if it showed up for you and you will be in the view where we can execute our queries. Note that if you are using the Fluid Azure App Insights logger, your telemetry data will be available in the “customEvents” table. 
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+1. Session information
+
+The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the document being interacted with, the number of collaborators and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; See the ...
```
```suggestion
The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the container being interacted with, the number of collaborators and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; see the intro paragraph for more information on sessions.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+2. Total number of sessions over time period
+
+The following query provides the average number of total sessions occurring over 10-minute intervals. The query logic graphs the total number of sessions occurring over 10-minute data points denoted by the variable named summedDataPointInterval. 
```
```suggestion
The following query provides the total number of sessions seen over 10-minute intervals. The query logic graphs the total number of sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. 
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+3. Average number of sessions over time period
+
+The following query provides the average number of total sessions occurring over 1-hour intervals. The query logic first sums up the total number of sessions occurring over 10-minute data points denoted by the variable named summedDataPointInterval. It then averages these data points over 1-hour intervals denoted by the averagedTimeInterval variable.
```
```suggestion
The following query provides the average number of sessions occurring over 1-hour intervals. The query logic first sums up the total number of sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. It then averages these data points over 1-hour intervals denoted by the `averagedTimeInterval` variable.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+4. Average number of containers per session over a time period
+
+The following query will provide you with the approximate average number of containers per sessions over 1 hour time intervals. The query logic first sums up the total number of collaborators per sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. It then averages these datapoints over 1-hour intervals denoted by the `averagedTimeInterval` variab ...
```
```suggestion
The following query will provide you with the approximate average number of containers per session over 1 hour time intervals. The query logic first sums up the total number of containers per sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. It then averages these datapoints over 1-hour intervals denoted by the `averagedTimeInterval` variable.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+5. Length of Individual Sessions in Minutes
+
+This query provides you with the length of time of individual sessions, graphed by putting them into time bins increasing in 2.5 minute increments. The query logic calculates the length of time of each session that occurs and then sorts them into time bins going up in 2.5 minute increments denoted by the `let sessionLengthMinuteBins = 2.5;` variable. To adjust the 2.5 minute time bins, simply mod ...
```
```suggestion
This query provides you with the length of time of individual sessions, graphed by putting them into time bins increasing in 2.5 minute increments. The query logic calculates the length of time of each session that occurs and then sorts them into time bins going up in 2.5 minute increments denoted by the `let sessionLengthMinuteBins = 2.5;` variable. To adjust the 2.5 minute time bins, simply modify the `sessionLengthMinuteBins` variable. Note that this number represents minutes.
```

alexvy86 commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -139,4 +138,183 @@ startTelemetry(telemetryConfig);
+
+2. Adjusting the gap of time that defines a session
+
+By default, we identify each session has a period of 5 minutes of inactivity before and after. However, you may want to adjust this time period. To do so, modify the value of the variable named sessionGap to be your desired time length.
```
```suggestion
By default, we identify each session has a period of 5 minutes of inactivity before and after. However, you may want to adjust this time period. To do so, modify the value of the variable named `sessionGap` to be your desired time length.
```

alexvy86 reviewed (approved):
Great docs. The comments are mostly typos, formatting, and little things like that. One or two questions among them.

rohandubal commented:
Huge props to Sean also for the major chunk of documentation internally which I used for conversion here!

rohandubal committed:
Apply suggestions from code review

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

rohandubal committed:
Add updates based on PR comment

rohandubal committed:
Merge branch 'main' into rohandubal/fluid-telemetry-docs-update

rohandubal committed:
Merge branch 'main' into rohandubal/fluid-telemetry-docs-update

seanimam commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -62,17 +73,25 @@ startTelemetry(telemetryConfig);
+That's it for now! If you've decided to use Azure App Insights, we have designed useful prebuilt queries that utilize the generated telemetry. You can try out these queries after running your application with above code for a little duration.
 
-## Example 2: Extending ITelemetryConsumer to send Fluid telemetry to a custom destination
+You can find these queries [in the "Interpreting Telemetry Data" section below.](#telemetry_visualization).
```
```suggestion
You can find these queries [in the "Interpreting Telemetry Data" section below.](#telemetry_visualization)
```
nit: extra period

seanimam commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -2,21 +2,29 @@
+### Prerequisite
 
-### Step 1: Install the @fluidframework/fluid-telemetry package and Azure App Insights package dependencies
+Before you can get telemetry sent to Azure App Insights, you'll need to create an Instance of App Insights on Azure. You'll then be able to leverage Azure App Insights integration and route your Fluid container application telemetry to App Insights. [Learn more about Azure App Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview). [Creating your App Insights i ...
```
```suggestion
Before you can get telemetry sent to Azure App Insights, you'll need to create an Instance of App Insights on Azure. You'll then be able to integrate your Azure App Insights instance with your Fluid application and route your Fluid container application telemetry to App Insights. [Learn more about Azure App Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview). [Creating your App Insights instance](https://learn.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource?tabs=bicep).
```

nit on wording, no strong feelings on this though.

seanimam commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -2,21 +2,29 @@
-#### Step 2: Now, let's start the telemetry production by initializing our telemetry collection where we initialize our containers:
+### Step 2: Setup telemetry collection
+
+Now, let's start the telemetry production by initializing our telemetry collection where we initialize our containers:
```
This was not part of your changes but on the example below on lines 32-33, we have two export lines from the same package that can be combined

seanimam commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -62,17 +73,25 @@ startTelemetry(telemetryConfig);
-### Step 2: First, we'll have to create our own telemetry consumer which extends the ITelemetryConsumer interface. Let's look at an example that will simply console.log the telemetry.
+### Step 2: Define a custom telemetry consumer
+
+Here we'll create our own telemetry consumer which extends the `ITelemetryConsumer` interface. Let's look at an example that will simply log the telemetry to console.
```
This was not part of your changes but on line 113 the comment should be changed to `// 1: import our implementation of MySimpleTelemetryConsumer from step 2` (right now it says step 1)

seanimam commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -116,3 +137,184 @@ startTelemetry(telemetryConfig);
+| extend Diff = datetime_diff("second", timestamp, prevTimestamp) 
+| extend IsNewPeriod = iif(prevContainerId != containerId or Diff > (sessionGapSeconds) or isnull(Diff), 1, 0) 
+| extend SessionId = row_cumsum(IsNewPeriod) 
+| summarize NumCollaborators = dcount(containerInstanceId), StartTime = min(timestamp), EndTime = max(timestamp) by SessionId, containerId 
+| extend PeriodDurationInMinutes = datetime_diff("minute", EndTime, StartTime) 
+| project SessionId, containerId , NumCollaborators, PeriodDurationInMinutes, StartTime, EndTime 
```
```suggestion
| summarize NumContainers = dcount(containerInstanceId), StartTime = min(timestamp), EndTime = max(timestamp) by SessionId, containerId 
| extend PeriodDurationInMinutes = datetime_diff("minute", EndTime, StartTime) 
| project SessionId, containerId , NumContainers, PeriodDurationInMinutes, StartTime, EndTime 
```

We might want to rename this from Collaborators to NumContainers, or just make it clear that its not like unique people. (FYI if you make this update you'll need an updated image. I can provide if needed)

seanimam replied:
![image](https://github.com/microsoft/FluidFramework/assets/105244057/1c5dfce5-e8e7-46e0-87e5-1b5731385aa6)

seanimam commented on packages/framework/client-logger/fluid-telemetry/README.md:
```diff
@@ -116,3 +137,184 @@ startTelemetry(telemetryConfig);
+
+1. Session information
+
+The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the container being interacted with, the number of collaborators and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; see th ...
```
We might want to adjust this wording to number of containers to avoid confusion. I have a new image if you choose to update the query in my suggestion below. We can also just clarify what we mean instead and keep the image/query the same if you like.
```suggestion
The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the container being interacted with, the number of containers and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; see the intro paragraph for more information on sessions.
```
![image](https://github.com/microsoft/FluidFramework/assets/105244057/1b243fa8-0f47-4fdc-a7ed-7ddfebb4317a)

rohandubal committed:
Update using
> @fluidframework/fluid-telemetry@2.0.0-rc.3.0.0 format /Users/rohandubal/Documents/Dev/FluidFramework/packages/framework/client-logger/fluid-telemetry
> fluid-build --task format .

Fluid Repo Root: /Users/rohandubal/Documents/Dev/FluidFramework
	@fluidframework/fluid-telemetry matched (.)
WARNING: @fluid-example/inventory-app: warning: dependent package fluid-framework not linked. Use --symlink or --symlink:full to fix.
WARNING: @fluidframework/server-services-shared: warning: dependent package @fluidframework/server-services-utils not linked. Use --symlink or --symlink:full to fix.
Symlink in isolated mode
Start tasks 'format' in 1 matched packages (1 total tasks in 34 packages)
[1/1] ✓ @fluidframework/fluid-telemetry: prettier --write . --cache --ignore-path ../../../../.prettierignore - 0.146s (non-incremental)
Build succeeded
Total time: 0.343s

rohandubal committed:
Apply suggestions from code review

Co-authored-by: seanimam <105244057+seanimam@users.noreply.github.com>

rohandubal committed:
Merge branch 'main' into rohandubal/fluid-telemetry-docs-update

rohandubal committed:
Merge branch 'main' into rohandubal/fluid-telemetry-docs-update
